### PR TITLE
Feat/python sdk storage

### PR DESCRIPTION
**Pull request by flopez7**
## Description

Rework storage module in python SDK, add proper documentation and tests (unit, integration).

## Summary of changes

- Added StorageClient class with constructor
- Added Credentials type
- Added download_files public method
- Added upload_files public method
- Added bucket_exists public method
- Added list_objects public method

## Related issues
#367 

## Operational checklist

- [x] All new functionality is covered by tests
- [x] Any related documentation has been changed or added

### CODE REVIEW

File: packages/sdk/python/human-protocol-sdk/test/human_protocol_sdk/storage/test_storage.py [hunk 1]
```diff
@@ -1,219 +1,242 @@
...
+        self.client.client.list_objects = MagicMock(side_effect=[[]])
+        result = self.client.list_objects(bucket=self.bucket)
+        self.assertEqual(result, [])
+
+    def test_list_objects_error(self):
+        self.client.client.head_bucket = MagicMock(
+            side_effect=Exception("Connection error")
+        )
+        with self.assertRaises(StorageClientError):
+            self.client.list_objects(bucket=self.bucket)
```
Comment (mrhouzlane) [inline packages/sdk/python/human-protocol-sdk/test/human_protocol_sdk/storage/test_storage.py:242]: ----

I added a test to check the length of objects passed is well returned. 
```

def test_list_objects_length(self):
        expected_length = random.randint(1, 10)
        mock_client = MagicMock()
        mock_client.list_objects.return_value = [types.SimpleNamespace(_object_name=f"file{i}") for i in range(expected_length)]
        with patch('human_protocol_sdk.storage.Minio', return_value=mock_client):
            client = StorageClient(endpoint_url='https://example.com', credentials=None)
            object_list = client.list_objects(bucket='my-bucket')
            self.assertEqual(len(object_list), expected_length)
            
    def test_list_objects_length_error(self):
        expected_length = random.randint(1, 10)
        mock_client = MagicMock()
        mock_client.list_objects.return_value = [types.SimpleNamespace(_object_name=f"file{i}") for i in range(expected_length)]
        with patch('human_protocol_sdk.storage.Minio', return_value=mock_client):
            client = StorageClient(endpoint_url='https://example.com', credentials=None)
            object_list = client.list_objects(bucket='my-bucket')
            if len(object_list) != expected_length:
                raise AssertionError(f"Expected {expected_length} objects, but found {len(object_list)}")

```

**Review by mrhouzlane — COMMENTED** [review]
Tests : 19 passed.
Added 2 : 21.

File: packages/sdk/python/human-protocol-sdk/human_protocol_sdk/job.py [hunk 2]
```diff
@@ -364,8 +374,18 @@ def launch(self, pub_key: bytes) -> bool:
         LOG.info("Job's escrow contract deployed to:{}".format(job_addr))
         self.job_contract = get_escrow(job_addr, self.hmt_server_addr)
 
-        (hash_, manifest_url) = upload(self.serialized_manifest, pub_key)
-        self.manifest_url = manifest_url
+        hash_ = self.storage.upload_files(
```
Comment (posix4e) [inline packages/sdk/python/human-protocol-sdk/human_protocol_sdk/job.py:377]: instead of hash_ let's call it the style of hash it is. Like is it a manifest hash? A input file hash?

File: packages/sdk/python/human-protocol-sdk/human_protocol_sdk/job.py [hunk 3]
```diff
@@ -651,16 +671,16 @@ def bulk_payout(
...
 
-        hash_, url = upload(
-            msg=results,
-            public_key=pub_key,
-            encrypt_data=encrypt_final_results,
-            use_public_bucket=store_pub_final_results,
+        hash_ = self.storage.upload_files([results], ESCROW_BUCKETNAME)[0]
+        url = (
+            ("https://" if ESCROW_SECURE else "http://")
+            + ESCROW_ENDPOINT_URL
```
Comment (posix4e) [inline packages/sdk/python/human-protocol-sdk/human_protocol_sdk/job.py:677]: URLS include https and stuff. Perhaps this is a hostname name a URL? Any reason why we don't just assume the url starts with http something or other, and add https:/// if it doesn't exist rather than an escrow secure variable?

File: packages/sdk/python/human-protocol-sdk/human_protocol_sdk/job.py [hunk 4]
```diff
@@ -27,16 +28,18 @@
...
-    get_public_bucket_url,
-    get_key_from_url,
-)
+from human_protocol_sdk.storage import StorageClient, Credentials
 
 GAS_LIMIT = int(os.getenv("GAS_LIMIT", 4712388))
 TRANSFER_EVENT = get_entity_topic(get_hmtoken_interface(), "Transfer")
 
+ESCROW_BUCKETNAME = os.getenv("ESCROW_PUBLIC_BUCKETNAME", "escrow-public-results")
+ESCROW_ACCESS_KEY = os.getenv("ESCROW_ACCESS_KEY", "minio")
```
Comment (posix4e) [inline packages/sdk/python/human-protocol-sdk/human_protocol_sdk/job.py:37]: Why is this being moved out of storage to job?

File: packages/sdk/python/human-protocol-sdk/human_protocol_sdk/storage.py [hunk 5]
```diff
@@ -1,238 +1,250 @@
...
+        result_files = []
+        for file in files:
+            try:
+                artifact = json.dumps(file, sort_keys=True)
+            except Exception as e:
+                LOG.error("Can't extract the json from the object")
+                raise e
+
+            data = artifact.encode("utf-8")
+            hash_ = hashlib.sha1(data).hexdigest()
```
Comment (posix4e) [inline packages/sdk/python/human-protocol-sdk/human_protocol_sdk/storage.py:172]: instead of hash_ lets call it something else like data_sha

File: packages/sdk/python/human-protocol-sdk/test/human_protocol_sdk/test_job.py [hunk 6]
```diff
@@ -387,22 +373,6 @@ def test_job_bulk_payout_with_full_qualified_url(self):
 
             transaction_retry_mock.reset_mock()
 
-            # Bulk payout with option to store final results publicly
```
Comment (posix4e) [inline packages/sdk/python/human-protocol-sdk/test/human_protocol_sdk/test_job.py:390]: Why was this deleted?

**Review by posix4e — DISMISSED** [review]
Mostly nits. I think we might be a bit short on testing now

File: packages/sdk/python/human-protocol-sdk/human_protocol_sdk/job.py (hunk 2 above)
Comment (flopez7) [reply packages/sdk/python/human-protocol-sdk/human_protocol_sdk/job.py:377]: @posix4e Everything related to job will be redone in a future PR. These are temporary fixes to make the old SDK version work

File: packages/sdk/python/human-protocol-sdk/test/human_protocol_sdk/storage/test_storage.py (hunk 1 above)
Comment (flopez7) [reply packages/sdk/python/human-protocol-sdk/test/human_protocol_sdk/storage/test_storage.py:242]: The list_objects function does not return the number of objects in the block, it directly returns a list of objects. Instead, there is another test that checks if the length and the values returned by the function are correct.

File: packages/sdk/python/human-protocol-sdk/test/human_protocol_sdk/test_job.py (hunk 6 above)
Comment (flopez7) [reply packages/sdk/python/human-protocol-sdk/test/human_protocol_sdk/test_job.py:390]: We have removed the option store_pub_final_results=True, so this test is exactly the same as the previous one. In addition, it expects to store the file publicly, but with the new module everything we store is publicly

**Review by Dzeranov — APPROVED** [review]
